### PR TITLE
[Security] Add migrating encoder configuration

### PR DESCRIPTION
**Pull request by chalasr**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        |

### CODE REVIEW

**Review by wouterj — APPROVED** [review]

File: src/Symfony/Bundle/SecurityBundle/CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,7 @@ CHANGELOG
 4.4.0
 -----
 
+ * Added `migrating_from` option to encoders configuration.
```
Comment (teohhanhui) [inline src/Symfony/Bundle/SecurityBundle/CHANGELOG.md]: Perhaps it should be named `migrate_from` (verb) for consistency with the other options, e.g. `ignore_case`.

File: src/Symfony/Component/Security/Core/Encoder/EncoderFactory.php [hunk 2]
```diff
@@ -175,4 +172,22 @@ private function getEncoderConfigFromAlgorithm(array $config): array
             ],
         ];
     }
+
+    private function createMigratingEncoder(array $config, array $algos)
+    {
+        foreach ($algos as $algo) {
+            // "plaintext" is not allowed as any leaked hashes could then be used to authenticate directly
+            if ('plaintext' === $algo) {
+                throw new LogicException('Migrating plaintext passwords is not supported.');
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Core/Encoder/EncoderFactory.php]: `is not allowed.`

**Review by nicolas-grekas — CHANGES_REQUESTED** [review]
I don't think the `migrating_from` option is needed at all: instead, we should *always* wire a `MigratingPasswordEncoder` when Sodium/Native encoder variants are selected.

There is one thing that is not documented correctly: when `auto` is selected, `hash_algorithm` is used as the digest algo too. But this should be fixed on 4.3, so it's unrelated to this PR.

File: src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/Fixtures/xml/migrating_encoder.xml [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+   xmlns:sec="http://symfony.com/schema/dic/security"
+   xsi:schemaLocation="http://symfony.com/schema/dic/services https://symfony.com/schema/dic/services/services-1.0.xsd">
+
+    <imports>
+        <import resource="container1.xml"/>
+    </imports>
+
+    <sec:config>
+        <sec:encoder class="JMS\FooBundle\Entity\User7" algorithm="argon2i" memory_cost="256" time_cost="1">
+            <sec:migrating_from>bcrypt</sec:migrating_from>
```
Comment (stof) [inline src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/Fixtures/xml/migrating_encoder.xml]: We use dashes in XML, not underscores

File: src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/Fixtures/xml/migrating_encoder.xml [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+   xmlns:sec="http://symfony.com/schema/dic/security"
+   xsi:schemaLocation="http://symfony.com/schema/dic/services https://symfony.com/schema/dic/services/services-1.0.xsd">
+
+    <imports>
+        <import resource="container1.xml"/>
+    </imports>
+
+    <sec:config>
+        <sec:encoder class="JMS\FooBundle\Entity\User7" algorithm="argon2i" memory_cost="256" time_cost="1">
```
Comment (stof) [inline src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/Fixtures/xml/migrating_encoder.xml]: should also use dashes

File: src/Symfony/Component/Security/Core/Encoder/EncoderFactory.php [hunk 5]
```diff
@@ -175,4 +172,22 @@ private function getEncoderConfigFromAlgorithm(array $config): array
...
+            }
+
+            $config['algorithm'] = $algo;
```
Comment (stof) [inline src/Symfony/Component/Security/Core/Encoder/EncoderFactory.php]: this assumes that all other settings are the same for all algos, which look wrong to me

**Review by stof — CHANGES_REQUESTED** [review]

**Comment by chalasr** [comment]
> There is one thing that is not documented correctly: when auto is selected, hash_algorithm is used as the digest algo too. But this should be fixed on 4.3, so it's unrelated to this PR.

That's where I disagree. Let's say one have `sha1` hashes in the user storage, this is the config required to make migrations work:
```yaml
App\User:
    algorithm: bcrypt
    hash_algorithm: md5
    encode_as_base64: false # also required because of the old encoder config
```

It's really not obvious what this means to me, it just feels like hijacking existing options to achieve what a new clearly-named option would do: migrate from a specific algo.

Hence I think we should rather improve the situation than documenting what currently kinda works by chance.
And I'd even say this PR is not enough yet: we need the ability to pass the required options per legacy algo you want to migrate from, so that things are clear just by looking at the config.

Built-in password migrations are great to make developers able to move forward and stop relying on old algos, I think that the support for this feature should be exhaustive and obvious for anyone reading the config, which is definitely not the case right now.

File: src/Symfony/Component/Security/Core/Encoder/EncoderFactory.php (hunk 5 above)
Comment (chalasr) [reply src/Symfony/Component/Security/Core/Encoder/EncoderFactory.php]: fixed

File: src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/Fixtures/xml/migrating_encoder.xml (hunk 4 above)
Comment (chalasr) [reply src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/Fixtures/xml/migrating_encoder.xml]: fixed

File: src/Symfony/Component/Security/Core/Encoder/EncoderFactory.php (hunk 2 above)
Comment (chalasr) [reply src/Symfony/Component/Security/Core/Encoder/EncoderFactory.php]: fixed

File: src/Symfony/Bundle/SecurityBundle/CHANGELOG.md (hunk 1 above)
Comment (chalasr) [reply src/Symfony/Bundle/SecurityBundle/CHANGELOG.md]: done!

**Comment by chalasr** [comment]
PR updated, the option now accepts encoder names in addition to algo names:
To migrate from an old encoder which defines some custom values for options, make it a named encoder (if it is not already) and pass its name to the `migrate_from` list. Same if it's a custom encoder service.

```yaml
encoders:
    foo:
         id: App\FooEncoder
    digest_sha256:
        algorithm: sha256
        encode_as_base64: false
     App\User:
         algorithm: argon2id
         migrate_from: 
             - foo
             - digest_sha256
             - bcrypt # creates an encoder with default config options
```

**Comment by chalasr** [comment]
@nicolas-grekas as discussed, sodium and native encoders are now always migrating.
PR ready

File: src/Symfony/Component/Security/Core/Encoder/EncoderFactory.php [hunk 6]
```diff
@@ -79,7 +80,23 @@ private function createEncoder(array $config): PasswordEncoderInterface
 
         $reflection = new \ReflectionClass($config['class']);
 
-        return $reflection->newInstanceArgs($config['arguments']);
+        $encoder = $reflection->newInstanceArgs($config['arguments']);
+
+        if (!$isExtra && \in_array($config['class'], [NativePasswordEncoder::class, SodiumPasswordEncoder::class], true)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Core/Encoder/EncoderFactory.php]: let's return here: `if ($isExtra || !\in_array...) return...`

File: src/Symfony/Component/Security/Core/Encoder/EncoderFactory.php [hunk 7]
```diff
@@ -89,7 +106,29 @@ private function getEncoderConfigFromAlgorithm(array $config): array
...
+            $encoderChain = [];
+            foreach ($fromEncoders as $name) {
+                if ($encoder = $this->encoders[$name] ?? false) {
+                    $encoder = $encoder instanceof PasswordEncoderInterface ? $encoder : $this->createEncoder($encoder, true);
+                } else {
+                    $encoder = $this->createEncoder(['algorithm' => $name], true);
+                }
+
+                if ($encoder instanceof PlaintextPasswordEncoder) {
+                    throw new LogicException('Migrating from plaintext encoders is not allowed.');
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Core/Encoder/EncoderFactory.php]: I'm not sure about this one: that's quite arbitrary - if one does want to migrate from plaintext, better not block them

Comment (chalasr) [reply src/Symfony/Component/Security/Core/Encoder/EncoderFactory.php]: removed

**Review by nicolas-grekas — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @chalasr.

**Pull request merged**